spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java

```diff
@@ -996,4 +1027,134 @@ public String unknown(
       return String.format("%s(%s) %s %s", transform, sourceName, direction, nullOrder);
     }
   }
+
+  /**
+   * Implement our own index in-memory index which will only list directories to avoid unnecessary
+   * file listings. Should ONLY be used to get partition directory paths. Uses table's schema to
+   * only visit partition dirs using number of partition columns depth recursively. Does NOT return
+   * files within leaf dir.
+   */
+  private static class InMemoryLeafDirOnlyIndex extends PartitioningAwareFileIndex {
```

Comment:
shall we add any test for this?

Reply:
yeah I guess I am just curious about the customization over `leafDirToChildrenFiles`, but have to admit there's no easy way to test this